### ISSUE TITLE
KeyError: 'pip.installed' -- after successful initial installation

### ISSUE THREAD

**Issue by myii** [saltstack-formulas/docker-formula#49]
Initial installation successful:

```
          ID: docker-py requirements
    Function: pip.installed
        Name: docker-py
      Result: True
     Comment: All packages were successfully installed
     Started: 22:33:56.745286
    Duration: 3161.768 ms
     Changes:
              ----------
              docker-py==1.4.0:
                  Installed
```

Encountered the following error when running `highstate` again:

```
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1161, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/state.py", line 523, in highstate
        whitelist=kwargs.get('whitelist')
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2968, in call_highstate
        return self.state.call_high(high)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 2067, in call_high
        ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1623, in call_chunks
        running = self.call_chunk(low, running, chunks)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1769, in call_chunk
        self._mod_init(low)
      File "/usr/lib/python2.7/dist-packages/salt/state.py", line 612, in _mod_init
        self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
      File "/usr/lib/python2.7/dist-packages/salt/utils/lazy.py", line 90, in __getitem__
        raise KeyError(key)
    KeyError: 'pip.installed'
```

Searched around and found that there have been various reports and suggestions about this error but no clear solution:

* https://github.com/saltstack/salt/issues/26493
* https://gist.github.com/nitay/a149724b5dffee95e8ac
* https://github.com/saltstack/salt/issues/26115
* https://github.com/saltstack/salt/pull/26160/files

Tried a few ideas and put together the following workaround using `pip.install` instead:

```
docker-py requirements:
  pkg.installed:
    - name: python-pip
  module.run:
    - name: pip.install
    {%- if "pip_version" in docker %}
    - pkgs: docker-py {{ docker.pip_version }}
    {%- else %}
    - pkgs: docker-py
    {%- endif %}
    - require:
      - pkg: docker package
    - reload_modules: True
```

That allowed the procedure to complete without errors:

```
          ID: docker-py requirements
    Function: module.run
        Name: pip.install
      Result: True
     Comment: Module function pip.install executed
     Started: 01:55:31.301498
    Duration: 441.122 ms
     Changes:
              ----------
              ret:
                  ----------
                  pid:
                      18652
                  retcode:
                      0
                  stderr:
                  stdout:
                      Requirement already satisfied (use --upgrade to upgrade): docker-py in /usr/local/lib/python2.7/dist-packages
                      Requirement already satisfied (use --upgrade to upgrade): requests>=2.5.2 in /usr/local/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): six>=1.4.0 in /usr/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): websocket-client>=0.32.0 in /usr/local/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): backports.ssl-match-hostname in /usr/local/lib/python2.7/dist-packages (from websocket-client>=0.32.0->docker-py)
                      Cleaning up...
```

Is there a better solution to this problem?

**Comment by timc3** [comment]
Getting what looks like the same problem:
```
----------
          ID: docker-py requirements
    Function: pip.installed
        Name: docker-py
      Result: False
     Comment: Python package docker-py was already installed
              Failed to install packages: . Error: You must give at least one requirement to install (see "pip help install")
              Storing complete log in /root/.pip/pip.log
     Started: 14:43:43.751117
    Duration: 289.68 ms
     Changes:
```

**Comment by myii** [comment]
@timc3 Above, I've described using `pip.install` as a workaround -- does that work for you, too?

**Comment by rallytime** [comment]
@myii That KeyError you're seeing is because of a regression in the state compiler in Salt. This is fixed already on 2015.8.0 and newer releases, and we're planning on doing another release on 2015.5 as well, so it will be fixed on 2015.5.7 once that is released.

**Comment by rallytime** [comment]
Also, as you've noticed, the workaround is to install the required dependency that is missing. That has fixed all of these types of stacktraces.

**Comment by chris-martin** [comment]
I tried @myii's workaround, and it gives this error:

```
local:
    Data failed to compile:
----------
    Rendering SLS 'base:docker' failed: Jinja error: argument of type 'StrictUndefined' is not iterable
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 306, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 6, in top-level template code
TypeError: argument of type 'StrictUndefined' is not iterable

; line 6

---
docker-py requirements:
  pkg.installed:
    - name: python-pip
  module.run:
    - name: pip.install
    {%- if "pip_version" in docker %}    <======================
    - pkgs: docker-py {{ docker.pip_version }}
    {%- else %}
    - pkgs: docker-py
    {%- endif %}
    - require:
[...]
---
```

**Comment by myii** [comment]
> rallytime commented 15 days ago
>
> @myii That KeyError you're seeing is because of a regression in the state compiler in Salt. This is fixed already on 2015.8.0 and newer releases, and we're planning on doing another release on 2015.5 as well, so it will be fixed on 2015.5.7 once that is released.

@rallytime Thanks for your feedback.  Still stuck on version `2015.5.3+ds-1trusty1` for the time being, I really need to get around to upgrading...

@chris-martin That `if` is part of the original source.  You could define `docker.pip_version` in your pillar.  Or simpler, if you want to work around it, just remove the `if` block.  Thus, use this instead:
```
docker-py requirements:
  pkg.installed:
    - name: python-pip
  module.run:
    - name: pip.install
    - pkgs: docker-py
    - require:
      - pkg: docker package
    - reload_modules: True
```

Or better still, as mentioned by @rallytime above, you could just upgrade and use the original source (haven't got around to doing this myself, yet).

As it is, I've just run `highstate` again and I'm having no problems:
```
----------
          ID: docker package dependencies
    Function: pkg.installed
      Result: True
     Comment: All specified packages are already installed.
     Started: 01:19:00.037085
    Duration: 1319.38 ms
     Changes:
----------
          ID: purge old packages
    Function: pkgrepo.absent
        Name: deb https://get.docker.com/ubuntu docker main
      Result: True
     Comment: Package repo deb https://get.docker.com/ubuntu docker main is absent
     Started: 01:19:01.357794
    Duration: 99.553 ms
     Changes:
----------
          ID: purge old packages
    Function: pkg.purged
        Name: lxc-docker*
      Result: True
     Comment: All specified packages are already absent.
     Started: 01:19:01.457431
    Duration: 5.731 ms
     Changes:
----------
          ID: docker package repository
    Function: pkgrepo.managed
        Name: deb https://apt.dockerproject.org/repo ubuntu-trusty main
      Result: True
     Comment: Package repo 'deb https://apt.dockerproject.org/repo ubuntu-trusty main' already configured
     Started: 01:19:01.463695
    Duration: 93.592 ms
     Changes:
----------
          ID: docker package
    Function: pkg.latest
        Name: docker-engine
      Result: True
     Comment: The following packages were successfully installed/upgraded: docker-engine
     Started: 01:19:01.557714
    Duration: 31303.317 ms
     Changes:
              ----------
              docker-engine:
                  ----------
                  new:
                      1.9.0-0~trusty
                  old:
                      1.8.3-0~trusty
----------
          ID: docker-config
    Function: file.managed
        Name: /etc/default/docker
      Result: True
     Comment: File /etc/default/docker is in the correct state
     Started: 01:19:32.867693
    Duration: 280.608 ms
     Changes:
----------
          ID: docker-service
    Function: service.running
        Name: docker
      Result: True
     Comment: Service docker is already enabled, and is in the desired state
     Started: 01:19:33.330546
    Duration: 247.636 ms
     Changes:
----------
          ID: docker-py requirements
    Function: pkg.installed
        Name: python-pip
      Result: True
     Comment: Package python-pip is already installed.
     Started: 01:19:33.580978
    Duration: 261.678 ms
     Changes:
----------
          ID: docker-py requirements
    Function: module.run
        Name: pip.install
      Result: True
     Comment: Module function pip.install executed
     Started: 01:19:33.844716
    Duration: 1232.787 ms
     Changes:
              ----------
              ret:
                  ----------
                  pid:
                      29341
                  retcode:
                      0
                  stderr:
                  stdout:
                      Requirement already satisfied (use --upgrade to upgrade): docker-py in /usr/local/lib/python2.7/dist-packages
                      Requirement already satisfied (use --upgrade to upgrade): requests>=2.5.2 in /usr/local/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): six>=1.4.0 in /usr/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): websocket-client>=0.32.0 in /usr/local/lib/python2.7/dist-packages (from docker-py)
                      Requirement already satisfied (use --upgrade to upgrade): backports.ssl-match-hostname in /usr/local/lib/python2.7/dist-packages (from websocket-client>=0.32.0->docker-py)
                      Cleaning up...

Summary
------------
Succeeded: 9 (changed=2)
Failed:    0
------------
Total states run:     9
```

In summary, getting rid off the `if` block should work for you.

**Comment by garrickp** [comment]
Just a quick +1, we're encountering this as well, even with a required install of the python-pip package. The module.run workaround, well, worked.

**Comment by myii** [comment]
@rallytime @garrickp With master and minions using `2015.8.3`, no longer have this problem so I'm going to close this issue -- thanks for your advice and feedback.

**Comment by rallytime** [comment]
@myii Glad that is working for you and thank you for following up here. :)

**Comment by myii** [comment]
@rallytime You're welcome.

Just a bit more feedback, in case anyone stumbles across this.  Since upgrading, I've now come across another `pip` problem with one of my minions, when attempting to install Docker using this formula:

```
An importable pip module is required but could not be found on your system. This usually means that the system's pip package is not installed properly.
```

Found some discussions about this:

* https://github.com/saltstack/salt/issues/28036
* https://github.com/saltstack/salt/issues/28239

Appears to be an issue with the package in the Ubuntu repos but I was surprised to encounter it with only one of my minions when all are running the same setup.  With a little investigation, noticed that `pip` was stuck on version `1.5.4` instead of being upgraded to `7.1.2`.

In my case, was able to resolve by upgrading `pip` manually:

```
$ sudo pip install --upgrade pip
```

Now able to use this Docker formula without problems.
